Add box resizing to MD integrators

clpetix opened the pull request.

clpetix committed:
Add volume scaling for Langevin ynamics.

clpetix commented:
@mphoward this is a working implementation of box resizing for LangevinDynamics. It is throwing a ton of errors on the unit tests, but I'm not sure how polished vs hacky we want this to be for now. 

Let me know how you'd like to proceed!

mphoward commented on src/relentless/simulate/md.py:
```diff
@@ -118,9 +118,10 @@ class RunLangevinDynamics(_Integrator):
     """
 
-    def __init__(self, steps, timestep, T, friction, seed, analyzers=None):
+    def __init__(self, steps, timestep, T, barostat, friction, seed, analyzers=None):
```
Can you add the `barostat=None` right before `analyzers` so it is an optional argument? Then, update and handle this accordingly in the HOOMD backend. That may fix some of your failing unit tests.

mphoward commented on src/relentless/simulate/hoomd.py:
```diff
@@ -709,6 +719,38 @@ def _call_v2(self, sim):
+                    yz = hoomd.variant.linear_interp([(0, yz1), (self.steps, yz2)])
+
+            hoomd.update.box_resize(
+                Lx=Lx, Ly=Ly, Lz=Lz, xy=xy, xz=xz, yz=yz, scale_particles=False
```
I would probably use `scale_particles=True` since that is also how a barostat works.

For this one that operates periodically, you also need to save it in a variable and disable & delete it after the run (see below)

mphoward commented on src/relentless/simulate/hoomd.py:
```diff
@@ -709,6 +719,38 @@ def _call_v2(self, sim):
 
+            if isinstance(self.barostat, extent.Extent):
+                if sim.dimension == 2:
+                    Lx, Ly, xy = self.barostat.as_array("HOOMD")
+                    Lz, xz, yz = 0.0, 0.0, 0.0
```
Unused arguments should be set to `None` so that the current value is not modified at all.

mphoward commented on src/relentless/simulate/hoomd.py:
```diff
@@ -709,6 +719,38 @@ def _call_v2(self, sim):
             for analyzer in self.analyzers:
                 analyzer.pre_run(sim, self)
 
+            if isinstance(self.barostat, extent.Extent):
```
In this case, the period of the updater should be set so that the action is taken immediately. I forget if that is `None` or something else.

clpetix replied:
I checked the 2.9.7 documentation and to start immediately then `phase=-1`. Also, `scale_particles` defaults to `True` so I have removed that argument.

mphoward replied:
I think you actually need `period=None` to have it act immediately (i.e., resize right away). Otherwise, the default period is 1, so it would try to update over time. Concurrently, you're also giving it a constant for the sizes, so these are probably essentially the same except that the updater will keep triggering over and over again.

So in summary, I think `period=None` when we just want to change the box size immediately (a single `Extent`), otherwise `period=1` to go between two boxes.

https://hoomd-blue.readthedocs.io/en/v2.9.7/module-hoomd-update.html#hoomd.update.box_resize

mphoward commented on src/relentless/simulate/hoomd.py:
```diff
@@ -709,6 +719,38 @@ def _call_v2(self, sim):
+                    Lx1, Ly1, Lz1, xy1, xz1, yz1 = self.barostat[0].as_array("HOOMD")
+                    Lx2, Ly2, Lz2, xy2, xz2, yz2 = self.barostat[1].as_array("HOOMD")
+
+                    Lx = hoomd.variant.linear_interp([(0, Lx1), (self.steps, Lx2)])
```
Should this be `self.steps-1` to make sure the right box size is achieved at the end of the integration? I think that is how HOOMD's variants work.

clpetix committed:
Update implementation to match HOOMD required syntax.

clpetix committed:
Update period to prevent triggering repeatedly.

clpetix committed:
Check if barostat is set before trying to scale box.

clpetix commented:
Found a small issue where there was no check on whether a barostat was set before applying the barostat logic. This caused an error when `barostat=None`. 

This seems to be running well in the test case discussed. Ready for a review whenever you get a chance next week, @mphoward !

mphoward reviewed (changes requested):
Thanks Levi, this looks great! I had some small comments.

In order to fix the issue with the unit tests, can you add the `barostat` argument to `RunLangevinDynamics` for the other simulation methods? If you put it in their constructors, they will be able to take it when they are constructed, and then they can just do nothing with it for the time being. I think that will allow everything to pass.

clpetix committed:
Add barostat to LAMMPS and Dilute simulation methods.

clpetix committed:
Remove default argument from backend operation.
Make barostat to be last argument.
Check that barostat has at most two Extents.

clpetix commented:
@mphoward I think this is all good and ready for use!

clpetix committed:
Set freud version less than 3

clpetix committed:
Require freud version 2.

clpetix committed:
Revert freud change.

clpetix committed:
Merge branch 'main' into feature/box-resize

clpetix committed:
Merge branch 'main' into feature/box-resize

mphoward committed:
Merge branch 'main' into feature/box-resize

clpetix committed:
Merge branch 'main' into feature/box-resize

clpetix committed:
Merge branch 'main' into feature/box-resize

clpetix committed:
Add  barostat  to RunBrownianDynamics initializer in md.py

clpetix committed:
Add barostat  to RunBrownianDynamics in dilute.py

clpetix committed:
Add box resizing functionality in HOOMD-blue.

clpetix committed:
Test box resizing in HOOMD.

clpetix committed:
Merge branch 'feature/box-resize' of https://github.com/mphowardlab/relentless into feature/box-resize

clpetix committed:
Merge branch 'main' of https://github.com/mphowardlab/relentless into feature/box-resize

clpetix committed:
Merge branch 'main' of https://github.com/mphowardlab/relentless into feature/box-resize

clpetix commented:
Making a note that we should ensure the the integrator steps is strictly an int. Else, we'll throw an error when using scientific notation.
